### ISSUE TITLE
Assumptions about project locking do not hold with load-after-store in WorkerExecutorParallelIntegrationTest

### ISSUE THREAD

**Issue by FrauBoes** [gradle/gradle#25540]
As part of the configuration-cache fix-it week in June 2023, the `WorkerExecutorParallelIntegrationTest` was reviewed but not fixed ([PR](https://github.com/gradle/gradle/pull/25464)). 

There are two test methods that make assumptions about project locking that do not hold with load-after-store:
- "does not start another task when a task awaits async work"
- "does not start task in another project when a task awaits async work"

The expectations in these two tests were changed as part of [this PR](https://github.com/gradle/gradle/pull/2171), from (1)
```
blockingHttpServer.expectConcurrent("task1-1", "task2")
blockingHttpServer.expectConcurrent("task1-2")
```
to (2)
```
blockingHttpServer.expectConcurrent("task1-1")
blockingHttpServer.expectConcurrent("task1-2")
blockingHttpServer.expectConcurrent("task2")
```

When running the tests with `configCacheIntegTest`, (1) passes and (2) fails with: 
```
at org.gradle.integtests.fixtures.executer.ForkingGradleHandle.unexpectedBuildStatus(ForkingGradleHandle.java:241)
	at org.gradle.integtests.fixtures.executer.ForkingGradleHandle.waitForStop(ForkingGradleHandle.java:223)
	at org.gradle.integtests.fixtures.executer.ForkingGradleHandle.waitForFinish(ForkingGradleHandle.java:191)
	at org.gradle.integtests.fixtures.executer.NoDaemonGradleExecuter.doRun(NoDaemonGradleExecuter.java:210)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.lambda$run$9(AbstractGradleExecuter.java:1208)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.run(AbstractGradleExecuter.java:1222)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.run(AbstractGradleExecuter.java:1207)
	at org.gradle.integtests.fixtures.executer.AbstractDelegatingGradleExecuter.doRun(AbstractDelegatingGradleExecuter.java:29)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.lambda$run$9(AbstractGradleExecuter.java:1208)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.run(AbstractGradleExecuter.java:1222)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.run(AbstractGradleExecuter.java:1207)
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.succeeds(AbstractIntegrationSpec.groovy:428)
	at org.gradle.workers.internal.WorkerExecutorParallelIntegrationTest.does not start another task when a task awaits async work(WorkerExecutorParallelIntegrationTest.groovy:738)

	at org.gradle.workers.internal.WorkerExecutorParallelIntegrationTest.does not start another task when a task awaits async work(WorkerExecutorParallelIntegrationTest.groovy:738)
	Suppressed: org.gradle.internal.exceptions.DefaultMultiCauseException: Failed to handle all HTTP requests.
		at app//org.gradle.test.fixtures.server.http.ChainingHttpHandler.waitForCompletion(ChainingHttpHandler.java:92)
		at app//org.gradle.test.fixtures.server.http.BlockingHttpServer.stop(BlockingHttpServer.java:380)
		at app//org.gradle.test.fixtures.server.http.BlockingHttpServer.after(BlockingHttpServer.java:405)
		at app//org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:59)
		at app//org.gradle.test.fixtures.file.AbstractTestDirectoryProvider$TestDirectoryCleaningStatement.evaluate(AbstractTestDirectoryProvider.java:101)
		at app//org.spockframework.junit4.AbstractRuleInterceptor.evaluateStatement(AbstractRuleInterceptor.java:66)
		at app//org.spockframework.junit4.TestRuleInterceptor.intercept(TestRuleInterceptor.java:43)
		at app//org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:101)
		at app//org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:156)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
		at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
		at java.base@11.0.18/java.util.ArrayList.forEach(ArrayList.java:1541)
		at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
		at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
		at app//org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:156)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
		at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
		at java.base@11.0.18/java.util.ArrayList.forEach(ArrayList.java:1541)
		at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
		at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
		at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
		at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
		at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
		at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
		at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
		at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
		at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
		at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
		at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
		at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
		at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
		at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
		at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
		at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
		at app//org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
		at app//com.gradle.enterprise.testdistribution.launcher.forked.LauncherMain$c.a(SourceFile:221)
		at app//com.gradle.enterprise.testdistribution.launcher.forked.LauncherMain$c.a(SourceFile:207)
		at app//com.gradle.enterprise.testdistribution.launcher.forked.LauncherMain$c.b(SourceFile:198)
		at app//com.gradle.enterprise.testdistribution.launcher.forked.LauncherMain$c.a(SourceFile:190)
		at app//com.gradle.enterprise.testdistribution.launcher.forked.LauncherMain.a(SourceFile:81)
		at app//com.gradle.enterprise.testdistribution.launcher.forked.LauncherMain.a(SourceFile:72)
		at app//com.gradle.enterprise.testdistribution.launcher.forked.l.a(SourceFile:20)
		at app//com.gradle.enterprise.testdistribution.launcher.forked.LauncherMain.main(SourceFile:57)
	Caused by: org.gradle.test.fixtures.server.http.UnexpectedRequestException: Unexpected request GET /task2 received. Waiting for 1 further requests, received [], released [], not yet received [GET /task1-1]
		at app//org.gradle.test.fixtures.server.http.UnexpectedRequestFailure.<init>(UnexpectedRequestFailure.java:25)
		at app//org.gradle.test.fixtures.server.http.ExpectationState.unexpectedRequest(ExpectationState.java:47)
		at app//org.gradle.test.fixtures.server.http.ExpectMaxNConcurrentRequests.selectResponseProducer(ExpectMaxNConcurrentRequests.java:121)
		at app//org.gradle.test.fixtures.server.http.ExpectMaxNConcurrentRequestsThenRelease.selectResponseProducer(ExpectMaxNConcurrentRequestsThenRelease.java:26)
		at app//org.gradle.test.fixtures.server.http.ChainingHttpHandler.selectProducer(ChainingHttpHandler.java:219)
		at app//org.gradle.test.fixtures.server.http.ChainingHttpHandler.handle(ChainingHttpHandler.java:136)
		at platform/jdk.httpserver@11.0.18/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
		at platform/jdk.httpserver@11.0.18/sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:82)
		at platform/jdk.httpserver@11.0.18/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:80)
		at platform/jdk.httpserver@11.0.18/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:848)
		at platform/jdk.httpserver@11.0.18/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
		at platform/jdk.httpserver@11.0.18/sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:817)
```

**Comment by ghale** [comment]
Ok, so the issue in those two tests is that work items task1-1 and task1-2 both execute inside the same task action.  It was important to not release the project lock inside a task action because we wanted to provide the guarantee that the project state would not change within the boundaries of a task action (even if you wait on async work to finish).  It can change _in between_ task actions, but not inside the boundary of a single task action.  And this is what these tests are checking.  So, this is why task1-1 and task2 would not execute at the same time - for "does not start another task when a task awaits async work" the two are in the same project and block on its lock; for "does not start task in another project when a task awaits async work" `--parallel` is not used, so there is only one "all projects" lock for the whole build.  Once we enable CC, the tasks can execute in parallel by default, so we see task1-1 and task2 execute concurrently.  Essentially, these tests are checking specifically that project locks are not released inside of a task action, but with CC, this isn't really a concern since the project is immutable anyways.  We _could_ change the tests in some way so that they work with CC, but then what would these tests even be validating?  It seems to me that these tests just don't apply when the project is immutable, so we just shouldn't execute them when CC is in play.

**Comment by FrauBoes** [comment]
Thanks for the additional context! So no action needed here, closing the issue.
